### DISCUSSION TITLE
Solaredge values to default energy dashboard help needed

### DISCUSSION THREAD

**Discussion by MikeTango** [WillCodeForCats/solaredge-modbus-multi#594] [Q&A]
Hi,

i noticed that possibly due to the way SE hands over data, some values in the native "energy" dashboard of HA are not plausible. 

Pictured: fail 1, it stacks the charged energy value and possibly adds the discharged energy value as a negative? I can only see one label that directly says which entity it comes from and i can't find the code for this panel / direct yaml editor.

![Bildschirmfoto_2024-05-01_10-06-16](https://github.com/WillCodeForCats/solaredge-modbus-multi/assets/3368005/72ea5d87-ee5b-45ec-90ac-1a4e5b906bea)

Pictured: fail 2, at the same time it seems to subtract the battery charge kwh from the production which is bull, possibly also the consumption. It is not showing an entity label that i could read, like M1.export which would be more plausible in this scenario. 

![Bildschirmfoto_2024-05-01_10-09-20](https://github.com/WillCodeForCats/solaredge-modbus-multi/assets/3368005/13a6b34d-0d88-4360-aad7-f4b06f06fbf9)

I guess i have to write a kwh entity that does the math i need? Help much appreciated.

**Comment by cyberkryten** [comment]
That's the way HA Engery dashboard works when you have batteries.  If you disable the battery then you will see the full 'Solar Production' values on the Energy Usage graph as a total of Solar Consumption and Export.

That also happens if you battery is 100% full and cannot take further charge - produced energy is consumed by usage of the house and export.

However, when some of the solar is going into a DC battery, it does not get shown on the Energy Usage tab as solar consumption by the house (because that isn't, it is going to the battery, not your devices in the house).  Instead it is shown as Battery Energy Import - this is the 'consumer' at that point and if it also showed it as 'Solar Consumption' you would have it double counted.

In your example picture 2, you had 2.3kW of Solar Production at 9am on the bottom graph, and it looks like that was pretty much all consumed by Export to the grid and charging the battery.

**Reply by Mougli** [reply]
Which entity did you use for solar production?

**Reply by cyberkryten** [reply]
Ah, I had some issues with that now I think about it.  If I recall correctly, as soon as I put the system into Remote Control mode it lost the internal sensor, so I had to come up with my own.

The problem with a Hybrid system is once that 'generation' sensor is gone, it is quite hard because you don't know from the AC power produced where it came from.

I have a whole variety of template sensors, from calculations to reading the optimisers.  All of them have issues/quirks!  Because the updates aren't always timed exactly you can end up with generation appearing overnight, so I added an 'is the sun up' check to remove that.

After tying loads of options, this is what I've settled on as the most accurate way of doing it on my system.

[Solar Generation Calc.txt](https://github.com/WillCodeForCats/solaredge-modbus-multi/files/15443136/Solar.Generation.Calc.txt)

**Reply by Mougli** [reply]
Thanks for sharing!

**Comment by SGXander** [comment]
Did you ever find a way to get the Solar production in as a viable figure for the energy dashboard?

I have the same issue as you I think: solaredge_i1_ac_energy is wrong when the solar power is used to charge the battery directly
I have a template sensor that almost fixes the issue: sensor.solar_production_energy which is defined as below;
```
{{(states("sensor.solaredge_i1_ac_energy") | float +
states("sensor.solaredge_b1_energy_import") | float) -
states("sensor.solaredge_b1_energy_export") | float}}
```
however this is wrong when the battery charges from the grid and I can't for the life of me figure out if there's a way around this with the available sensors. Why is the solar production figure not exported via modbus :(

Edit: Solved. See comments

**Reply by cyberkryten** [reply]
Yeah, I figured that it doesn't allow for grid charging during solar production times.  

Removing the InvDV < 0 check in my example sorts that but has other issues.

Without that check, it means you get phantom spikes when charging or discharging starts/stops, I have 10kW of panels (E/W) and at 5-6am often see 12kW of 'generation' due to mismatched figures.

I'm working on an updated template that takes into account both the inverter max charge rate and battery states, will post if I get something that works reliably but having three batteries complicates things a lot as there are times when two are charging and one discharging for a brief second.

**Reply by SGXander** [reply]
Ok so forgive me for being dense but doesnt the sun up detection only apply to current power i.e. Watts? How can you apply it to the kWh energy readings without making them fluctuate wildly?

**Reply by cyberkryten** [reply]
Sorry, what I didn't say is that I have a Helper setup as an integral to calculate the Energy (kWh) from the power values.  This is pretty accurate as the modbus sampling is fast enough that any slight delay isn't noticeable in the final energy readings.

  Name: Calc SolarEdgeModbus Solar Generation Energy Alt
  Input Sensor: SEModbus Power - Solar Generation Alt
  Integration method: Trapezoidal rule
  Precision: 2

So, because I'm using the energy calculated from the power sensor, it means if the sun isn't up, power will be zero which means energy will also be zero.  

For Solar Panels on the Energy tab I use the "Calc SolarEdgeModbus Solar Generation Energy Alt" sensor
Then I add each battery individually and it tracks the state of all three:
  Solaredgemodbus B1 Energy Import
  Solaredgemodbus B1 Energy Export

**Reply by SGXander** [reply]
OK that's what I'm misisng out on then! I have a template for the power sensor already that seems to be fairly accurate so I'll have a crack at the integral later. What is your modbus frequency set to? I'm 30s at the moment.

**Reply by SGXander** [reply]
That does indeed work a treat. Getting good data now and it matches the solaredge app within a few 10s of wh.

For ease of anyone looking for this (This is for a single-inverter + single-battery setup with default entity names):
First create a template helper with the following settings;
- Name: Solar Production Power
- State Template:
```
{% set inverterpower = states("sensor.solaredge_i1_ac_power") | float %}
{% set batterypower = states("sensor.solaredge_b1_dc_power") | float %}
{% if inverterpower > 0 or batterypower > 0 %}
  {{max(0, (inverterpower + batterypower))}}
{% else %}
  0
{% endif %}
```
- Unit Of Measurement: W
- Device Class: Power
- State Class: Measurement
- Device: Solaredge I1

Next create an integral helper with the following settings;
- Name: Solar Production Energy
- Metric Prefix: k (kilo)
- Time Unit: Hours
- Input Sensor: Solar Production Power
- Integration Method: Trapezoidal rule
- Precision: 2
- Max sub-interval: 0:00:00

Lastly go to Settings > Dashboards > Energy and replace the solar entity with sensor.solar_production_energy

**Reply by cyberkryten** [reply]
Glad it was useful!

I started with something similar but such a basic config causes me all sorts of issues - does that template show that there is any generation at night or major spikes/troughs that coincide with large load changes?  

I'm wondering if that only occurs with multiple batteries.  I had got to a template of nearly 150 lines to try to stop all the ridiculous readings (eg 15kW of Solar at 2am!).  

I'm now looking at a different approach as the whack-a-mole of fixing each oddity just hasn't worked.

**Reply by SGXander** [reply]
15kW at night sounds odd. I haven't seen any night-generation yet but was thinking you could put in another section to the if statement to check the inverters mode is production. I think when in idle or other modes it draws a little bit but should be nowhere near 15kW!

can give the simple method a try with different names and see how they stack up over a couple of days before replacing though I'd be keen to know the difference between single and multi in the config.

**Comment by cyberkryten** [comment]
Just noticed you're subtracting the battery DC power from the Inverter AC power - is there a reason you aren't using the Inverter DC power?  sensor.solaredge_i1_dc_power

If you aren't getting spikes then I'm guessing it's an issue with multiple batteries only.  When the inverter changes from charge to discharge, that shows on the _i1_ values and then a second or so later on the _b1/2/3_ values - that timing difference is what causes the issue.

Yellow here is the raw values with no smoothing or checked for sun being 'up'.  

Orange is the smoothed version with change limited to +/- 5%

<img width="1671" alt="Solar base v smoothed" src="https://github.com/user-attachments/assets/56724330-c323-478d-b095-c74191462563">

**Reply by SGXander** [reply]
That's awesome, how did you do the change limit for the orange?
Re the AC I couldn't spot any difference in the values whenever checking, the DC always seems to be within 0.5w of AC but it's var'd so easy to change.

**Reply by cyberkryten** [reply]
The calculations to figure out what power is coming from the Solar panels in a Hybrid system are difficult and often inaccurate because of differences in sampling time between the Inverter DC figure and the Battery DC figures.  When more than one battery is involved (I have three) then these problems increase as there are times when the system reports that two are charging and one is discharging, giving odd totals.

Calculating correct Solar is especially difficult at times of significant increase or decrease in load or when transitions occur between battery discharge to load, export to grid and battery charging from grid.

At these times, the difference in DC power levels is often many kW (it is +/- 8kW in my system for a 16kW swing) and this results in huge spikess/troughs in the apparent 'Solar Generation' calculated from the DC values available when they do not line up exactly.

I've repeatedly added code to spot and account for these edge cases as and when they have occurred, to the point that my template for Solar Generation reached 150 lines and was still having issues with more edge cases.

As a result, I've found that method of trying to calculate around all of these different possibilities is pointless and have instead settled on a new method of generating a more accurate figure.

Solar changes are mostly relatively slow, with wild swings of many kW unlikely.  However, when clouds cover or uncover the sun there are some cases where large swings occur.  
If the Batteries are full then there's no issue and those figures can be used, but if the batteries are charging, it is impossible to determine if a swing is due to solar increase, timing issues or load changes (especially noticeable after clipping as there is latent power ready to be produced)

So, what I've done is limit the increase/decrease of solar generation to a percentage of the previous value.  This has it's issues, as it will take time for values to increase or decay, but overall it seems more accurate and less prone to major errors than my previous method.

There's some adjustments needed in this approach as when dealing with very low values, it is hard to get an increase/decrease of significance using a small percentage, so I've defined a low point at which values can go up/down by a larger amount.

The biggest issue is that when generation changes a lot, the values lag behind the real ones, especially when generation is low.  When under 1kW it can take a while to ramp up/down with the 5% setting, but this is only 2-3min in most cases.  More frequent polling (and therefore more updates) will reduce this (so going from 30sec to 15sec Modbus refresh makes a big difference)

Generation is allowed from dawn (before sunrise) until sunset.  There's some useful accumulation of solar energy in the hour or so before official sunrise, hence why I've changed from using sunrise as the start point.  Without limiting this, I see 'generation' overnight due to the differences in DC between batteries and inverter.

The template also has specific triggers so it only fires on change in the total DC values or every minute.

Settings:
Requires the 'Sun' integration to determine sun elevation and rising state
BattDC: Set to the sensor you are using which calculates the total DC figure for all batteries.
InvDC: Set to the sensor which provides the DC figure for the Inverter 

MaxPanels: Generation is capped at the Max rated for all the panels installed - set this to your total.  In my case, I'll never reach it (E/W split) but it is a useful sanity check to avoid errors giving any excessive values.

LowVal: This is the number of W the figure is allowed to jump to from zero.  Lower is better but very low will cause a very slow ramp up and increased lag at sunrise.
It is also used to decide at what point (one tenth of LowVal) the generation value is allowed to drop straight to zero - otherwise, you'll never reach zero due to rounding.

PCTInc/PCTDec: Percentage change for increase/decrease of the values.  I suspect this will depend on your installation, panels and a lot of other factors.  Bigger values will allow less lag, but will also produce larger spikes/troughs from the 'known problem' events.


Attached at .TXT as .YAML files aren't allowed:

[Template-SolarGenerationYaml.txt](https://github.com/user-attachments/files/16619619/Template-SolarGenerationYaml.txt)
